### PR TITLE
introduced concept of stream guard to throw an exception on invalid s…

### PR DESCRIPTION
**Pull request by kodjo-anipah**
Added stream deletion guard concept so that we can check in the enterprise plugin whether the stream should be allowed for deletion or not.

/nocl
/prd https://github.com/Graylog2/graylog-plugin-enterprise/pull/7260

### CODE REVIEW

**Comment by bernd** [comment]
@kodjo-anipah We already have the "scoped database entities" concept. Can we use it for your use case?

File: graylog2-server/src/main/java/org/graylog2/rest/resources/streams/DefaultStreamDeletionGuard.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * Server Side Public License for more details.
+ *
+ * You should have received a copy of the Server Side Public License
+ * along with this program. If not, see
+ * <http://www.mongodb.com/licensing/server-side-public-license>.
+ */
+package org.graylog2.rest.resources.streams;
```
Comment (bernd) [inline graylog2-server/src/main/java/org/graylog2/rest/resources/streams/DefaultStreamDeletionGuard.java:17]: We should put this into the `org.graylog2.streams` package. The guard should not only protect requests via the HTTP API.

File: graylog2-server/src/main/java/org/graylog2/rest/resources/streams/DefaultStreamDeletionGuard.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+import jakarta.inject.Singleton;
+
+@Singleton
+public class DefaultStreamDeletionGuard implements StreamDeletionGuard {
+
+    @Override
+    public void checkGuard(String streamId) {
+        // Do nothing in graylog open
```
Comment (bernd) [inline graylog2-server/src/main/java/org/graylog2/rest/resources/streams/DefaultStreamDeletionGuard.java:26]: Why do you add a default guard if it doesn't do anything?

File: graylog2-server/src/main/java/org/graylog2/rest/resources/streams/StreamResource.java [hunk 3]
```diff
@@ -378,6 +381,13 @@ public void delete(@ApiParam(name = "streamId", required = true) @PathParam("str
     private Stream deleteInner(String streamId, UserContext userContext) throws NotFoundException {
         checkPermission(RestPermissions.STREAMS_EDIT, streamId);
         checkNotEditableStream(streamId, "The stream cannot be deleted.");
+        streamDeletionGuards.forEach(guard -> {
```
Comment (bernd) [inline graylog2-server/src/main/java/org/graylog2/rest/resources/streams/StreamResource.java:384]: Let's move the stream guard checks into the `StreamService` to ensure we apply the guards for all stream deletions, not only the deletions via HTTP API.

**Review by bernd — CHANGES_REQUESTED** [review]

File: graylog2-server/src/main/java/org/graylog2/rest/resources/streams/DefaultStreamDeletionGuard.java (hunk 2 above)
Comment (kodjo-anipah) [reply graylog2-server/src/main/java/org/graylog2/rest/resources/streams/DefaultStreamDeletionGuard.java:26]: you are right, just declaring the Multibinder is enough, fixing that

File: graylog2-server/src/main/java/org/graylog2/contentpacks/facades/StreamFacade.java [hunk 4]
```diff
@@ -268,8 +269,12 @@ public void delete(Stream nativeEntity) {
             return;
         }
         try {
+            LOG.debug("WTF");
```
Comment (bernd) [inline graylog2-server/src/main/java/org/graylog2/contentpacks/facades/StreamFacade.java:272]: Is this a debugging leftover or should we add a real debug message?

File: graylog2-server/src/main/java/org/graylog2/database/PersistedServiceImpl.java [hunk 5]
```diff
@@ -236,13 +236,13 @@ public Map<String, List<ValidationResult>> validate(Map<String, Validator> valid
...
                 if (validationResult instanceof ValidationResult.ValidationFailed) {
-                    LOG.debug("Validation failure: [{}] on field [{}]", v.getClass().getCanonicalName(), field);
+                    LOG.info("Validation failure: [{}] on field [{}]", v.getClass().getCanonicalName(), field);
                     validationErrors.computeIfAbsent(field, k -> new ArrayList<>());
                     validationErrors.get(field).add(validationResult);
                 }
             } catch (Exception e) {
                 final String error = "Error while trying to validate <" + field + ">, got exception: " + e;
-                LOG.debug(error);
+                LOG.info(error);
```
Comment (bernd) [inline graylog2-server/src/main/java/org/graylog2/database/PersistedServiceImpl.java:245]: Are these log messages supposed to stay on INFO?

File: graylog2-server/src/main/java/org/graylog2/plugin/inject/Graylog2Module.java [hunk 6]
```diff
@@ -56,6 +56,7 @@
 import org.graylog2.plugin.security.PasswordAlgorithm;
 import org.graylog2.plugin.security.PluginPermissions;
 import org.graylog2.plugin.validate.ClusterConfigValidator;
+import org.graylog2.streams.StreamDeletionGuard;
```
Comment (bernd) [inline graylog2-server/src/main/java/org/graylog2/plugin/inject/Graylog2Module.java:59]: This adds an unused import to the file.

**Review by bernd — CHANGES_REQUESTED** [review]

File: graylog2-server/src/main/java/org/graylog2/contentpacks/facades/StreamFacade.java (hunk 4 above)
Comment (kodjo-anipah) [reply graylog2-server/src/main/java/org/graylog2/contentpacks/facades/StreamFacade.java]: @bernd  sorry about that I forgot to push, my bad

**Review by bernd — APPROVED** [review]
Thank you for the fixes! :pray: The backend looks good now.

**Pull request merged**